stdlib/src/memory/arc.mojo

```diff
@@ -45,32 +45,231 @@ print(Arc(String("ok"))[])
+        @parameter
+        if enable_weak:
+            return self.drop_weak()
+        else:
+            return last_strong
+
+    fn drop_weak(inout self) -> Bool:
+        """
+        Atomically decrement the weakref count, and return
+        true if the result hits zero
+        Returns:
+            True iff the backing allocation for `self`
+            should (must!) be deleted
+        """
+        @parameter
+        if enable_weak:
+            var last_any = self._weakref_ptr()[].fetch_sub(1) == 1
+
+            # if this is the last weak, and there
+            # are no other strongs, then we can dealloc the inner
+            return last_any
+        else:
+            #TODO: statically ensure unreachable?
+            return False
+
+
+@register_passable
+struct Weak[T: Movable, enable_weak: Bool = True](CollectionElement, CollectionElementNew):
+    """A Weak[T] is used much like an Arc[T], except it does not
+    prevent the target instance of T from being deinitialized/deallocated.
+
+    This is useful for data structures like cyclic graphs and doubly linked lists,
+    such that users can easily avoid memory leaks.
+
+    Parameters:
+        T: The type of the value that this Weak may be upgraded into an Arc[T] of.
+    """
+
+    alias _inner_type = _ArcInner[T]
+    var _inner: UnsafePointer[Self._inner_type]
+
+    fn __init__(inout self, strong: Arc[T, enable_weak]):
+        """Allows creating a Weak[T] given an Arc[T].
+
+        Args:
+            strong: The Arc[T] that we want a new Weak for the value of.
+        """
+        constrained[enable_weak, "Creating a Weak pointer from an `Arc[T, enable_weak = False]` is impossible." + 
+            "In order to create an Arc[T] that allows `downgrade()`, change the `enable_weak` type parameter to `True`."]()
+
+        strong._inner[].add_weak()
+        self._inner = strong._inner
+
+    fn __init__(inout self, other: Self):
+        """Copy constructor, produces a new Weak that is a copy of another Weak.
+
+        Args:
+            other: The other Weak[T] this should be a copy of.
+        """
+        other._inner[].add_weak()
+        self._inner = other._inner
+
+    @no_inline
+    fn __del__(owned self):
+        """Destructor for Weak[T], drops the weak ref and frees the backing
+        storage for the Inner iff this is the last reference of any sort to the inner.
+        """
+        var should_del = self._inner[].drop_weak()
+        if should_del:
+            (self._inner).destroy_pointee()
+            self._inner.free()
+        self._inner = UnsafePointer[Self._inner_type]()
+
+    fn __copyinit__(inout self, other: Self):
+        """Implicit copy of Weak[T],
+        simply delegates to the regular explicit copy constructor.
+
+        Args:
+            other: The Weak[T] to (implicitly) copy.
+        """
+        other._inner[].add_weak()
+        self._inner = other._inner
+
+    fn upgrade(inout self) -> Optional[Arc[T]]:
```

Comment:
**Suggestion** I think this needs to encode that this is an `Arc` containing a weak count:

```suggestion
    fn upgrade(inout self) -> Optional[Arc[T, True]]:
```